File: backend/app/api/phase4.py
"""Phase 4 QA assistant API routes."""

from typing import NoReturn

from fastapi import APIRouter, Depends, HTTPException, status

from backend.app.api.deps import RequestContext, get_request_context
from backend.app.api.rate_limit import ai_rate_limit
from backend.app.schemas.phase4 import (
    BugReportResponse,
    BugReportsResponse,
    EligibleStoriesResponse,
    FailGateRequest,
    FixLogResponse,
    GenerateBugReportRequest,
    GenerateBugReportResponse,
    GenerateEdgeCasesRequest,
    GenerateEdgeCasesResponse,
    GenerateTestPlanRequest,
    GenerateTestPlanResponse,
    PassGateRequest,
    SaveBugReportRequest,
    StoryContextResponse,
    SaveTestPlanRequest,
    TestPlanResponse,
    TestPlansResponse,
)
from backend.app.schemas.workspace import OkResponse
from backend.app.services.phase4_service import Phase4Service, Phase4ValidationError
from src.ai_engine import AIError, AIRateLimitError, AITimeoutError

router = APIRouter()


def get_phase4_service() -> Phase4Service:
    return Phase4Service()


def _handle_error(exc: Exception) -> NoReturn:
    if isinstance(exc, Phase4ValidationError):
        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_CONTENT, detail=str(exc)) from exc
    if isinstance(exc, AIRateLimitError):
        raise HTTPException(status_code=status.HTTP_429_TOO_MANY_REQUESTS, detail=str(exc)) from exc
    if isinstance(exc, AITimeoutError):
        raise HTTPException(status_code=status.HTTP_504_GATEWAY_TIMEOUT, detail=str(exc)) from exc
    if isinstance(exc, AIError):
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=str(exc)) from exc
    if isinstance(exc, EnvironmentError):
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=str(exc)) from exc
    raise exc


@router.get("/eligible-stories", response_model=EligibleStoriesResponse)
def eligible_stories(
    ctx: RequestContext = Depends(get_request_context),
    service: Phase4Service = Depends(get_phase4_service),
):
    try:
        return {"stories": service.get_eligible_stories(ctx)}
    except Exception as exc:
        _handle_error(exc)


@router.get("/story-context/{story_id}", response_model=StoryContextResponse)
def story_context(
    story_id: int,
    ctx: RequestContext = Depends(get_request_context),
    service: Phase4Service = Depends(get_phase4_service),
):
    try:
        return service.get_story_context(ctx, story_id)
    except Exception as exc:
        _handle_error(exc)


@router.post("/generate-test-plan", response_model=GenerateTestPlanResponse)
def generate_test_plan(
    payload: GenerateTestPlanRequest,
    ctx: RequestContext = Depends(get_request_context),
    service: Phase4Service = Depends(get_phase4_service),
    _rl: None = Depends(ai_rate_limit),
):
    try:
        extra_kwargs = {"extra_context_files": payload.extra_context_files} if payload.extra_context_files else {}
        md = service.generate_test_plan(
            ctx, payload.story_id, payload.instructions, payload.emphasis,
            **extra_kwargs,
        )
        return {"story_id": payload.story_id, "test_plan_md": md}
    except Exception as exc:
        _handle_error(exc)


@router.post("/generate-edge-cases", response_model=GenerateEdgeCasesResponse)
def generate_edge_cases(
    payload: GenerateEdgeCasesRequest,
    ctx: RequestContext = Depends(get_request_context),
    service: Phase4Service = Depends(get_phase4_service),
    _rl: None = Depends(ai_rate_limit),
):
    try:
        extra_kwargs = {"extra_context_files": payload.extra_context_files} if payload.extra_context_files else {}
        md = service.generate_edge_cases(
            ctx, payload.story_id, payload.scenario_text,
            **extra_kwargs,
        )
        return {"story_id": payload.story_id, "edge_cases_md": md}
    except Exception as exc:
        _handle_error(exc)


@router.post("/save-test-plan", response_model=OkResponse)
def save_test_plan(
    payload: SaveTestPlanRequest,
    ctx: RequestContext = Depends(get_request_context),
    service: Phase4Service = Depends(get_phase4_service),
):
    try:
        service.save_test_plan(ctx, payload.story_id, payload.test_plan_md)
        return {"ok": True}
    except Exception as exc:
        _handle_error(exc)


@router.get("/test-plans", response_model=TestPlansResponse)
def list_test_plans(
    ctx: RequestContext = Depends(get_request_context),
    service: Phase4Service = Depends(get_phase4_service),
):
    try:
        return {"test_plans": service.list_all_test_plans(ctx)}
    except Exception as exc:
        _handle_error(exc)


@router.get("/test-plan/{story_id}", response_model=TestPlanResponse)
def get_test_plan(
    story_id: int,
    ctx: RequestContext = Depends(get_request_context),
    service: Phase4Service = Depends(get_phase4_service),
):
    try:
        md = service.load_test_plan(ctx, story_id)
        return {"story_id": story_id, "test_plan_md": md}
    except Exception as exc:
        _handle_error(exc)


@router.delete("/test-plan/{story_id}", response_model=OkResponse)
def delete_test_plan(
    story_id: int,
    ctx: RequestContext = Depends(get_request_context),
    service: Phase4Service = Depends(get_phase4_service),
):
    try:
        service.delete_test_plan(ctx, story_id)
        return {"ok": True}
    except Exception as exc:
        _handle_error(exc)


@router.post("/generate-bug-report", response_model=GenerateBugReportResponse)
def generate_bug_report(
    payload: GenerateBugReportRequest,
    ctx: RequestContext = Depends(get_request_context),
    service: Phase4Service = Depends(get_phase4_service),
    _rl: None = Depends(ai_rate_limit),
):
    try:
        md = service.generate_bug_report(
            ctx,
            payload.story_id,
            [fs.model_dump() for fs in payload.failed_scenarios],
        )
        return {"story_id": payload.story_id, "bug_report_md": md}
    except Exception as exc:
        _handle_error(exc)


@router.post("/pass-gate", response_model=OkResponse)
def pass_gate(
    payload: PassGateRequest,
    ctx: RequestContext = Depends(get_request_context),
    service: Phase4Service = Depends(get_phase4_service),
):
    try:
        service.pass_gate(
            ctx,
            payload.story_id,
            scenario_results=(
                [r.model_dump() for r in payload.scenario_results]
                if payload.scenario_results else None
            ),
        )
        return {"ok": True}
    except Exception as exc:
        _handle_error(exc)


@router.post("/fail-gate", response_model=OkResponse)
def fail_gate(
    payload: FailGateRequest,
    ctx: RequestContext = Depends(get_request_context),
    service: Phase4Service = Depends(get_phase4_service),
):
    try:
        service.fail_gate(
            ctx,
            payload.story_id,
            payload.bug_report_md,
            payload.root_cause,
            payload.resolution_summary,
            scenario_results=(
                [r.model_dump() for r in payload.scenario_results]
                if payload.scenario_results else None
            ),
        )
        return {"ok": True}
    except Exception as exc:
        _handle_error(exc)


@router.get("/bug-reports", response_model=BugReportsResponse)
def list_bug_reports(
    ctx: RequestContext = Depends(get_request_context),
    service: Phase4Service = Depends(get_phase4_service),
):
    try:
        return {"bug_reports": service.list_all_bug_reports(ctx)}
    except Exception as exc:
        _handle_error(exc)


@router.get("/fix-log", response_model=FixLogResponse)
def fix_log(
    ctx: RequestContext = Depends(get_request_context),
    service: Phase4Service = Depends(get_phase4_service),
):
    try:
        return {"fix_log_md": service.get_fix_log(ctx)}
    except Exception as exc:
        _handle_error(exc)


@router.get("/bug-report/{story_id}", response_model=BugReportResponse)
def get_bug_report(
    story_id: int,
    ctx: RequestContext = Depends(get_request_context),
    service: Phase4Service = Depends(get_phase4_service),
):
    try:
        return {"story_id": story_id, "bug_report_md": service.load_bug_report(ctx, story_id)}
    except Exception as exc:
        _handle_error(exc)


@router.post("/save-bug-report", response_model=OkResponse)
def save_bug_report(
    payload: SaveBugReportRequest,
    ctx: RequestContext = Depends(get_request_context),
    service: Phase4Service = Depends(get_phase4_service),
):
    try:
        service.save_bug_report(ctx, payload.story_id, payload.bug_report_md)
        return {"ok": True}
    except Exception as exc:
        _handle_error(exc)


@router.delete("/bug-report/{story_id}", response_model=OkResponse)
def delete_bug_report(
    story_id: int,
    ctx: RequestContext = Depends(get_request_context),
    service: Phase4Service = Depends(get_phase4_service),
):
    try:
        service.delete_bug_report(ctx, story_id)
        return {"ok": True}
    except Exception as exc:
        _handle_error(exc)

File: frontend/lib/api/phase4.ts
import { apiRequest } from "./client";
import type {
  Phase4BugReportResponse,
  Phase4BugReportsResponse,
  Phase4EligibleStoriesResponse,
  Phase4FailGateRequest,
  Phase4FixLogResponse,
  Phase4GenerateBugReportRequest,
  Phase4GenerateBugReportResponse,
  Phase4GenerateTestPlanResponse,
  Phase4ScenarioResultItem,
  Phase4StoryContext,
  Phase4TestPlanResponse,
  Phase4TestPlansResponse,
  RequestContext,
} from "./types";

export const PHASE4_AI_TIMEOUT_MS = 480_000;

export function getEligibleStories(context: RequestContext) {
  return apiRequest<Phase4EligibleStoriesResponse>("/api/phase4/eligible-stories", { context });
}

export function getStoryContext(context: RequestContext, storyId: number) {
  return apiRequest<Phase4StoryContext>(`/api/phase4/story-context/${storyId}`, { context });
}

export function generateTestPlan(context: RequestContext, storyId: number, signal?: AbortSignal, instructions?: string) {
  return apiRequest<Phase4GenerateTestPlanResponse>("/api/phase4/generate-test-plan", {
    method: "POST",
    context,
    body: { story_id: storyId, ...(instructions?.trim() ? { instructions } : {}) },
    timeoutMs: PHASE4_AI_TIMEOUT_MS,
    signal,
  });
}

export function generateEdgeCases(context: RequestContext, storyId: number, scenarioText: string, signal?: AbortSignal) {
  return apiRequest<{ story_id: number; edge_cases_md: string }>("/api/phase4/generate-edge-cases", {
    method: "POST",
    context,
    body: { story_id: storyId, scenario_text: scenarioText },
    timeoutMs: PHASE4_AI_TIMEOUT_MS,
    signal,
  });
}

export function saveTestPlan(context: RequestContext, storyId: number, testPlanMd: string) {
  return apiRequest<{ ok: boolean }>("/api/phase4/save-test-plan", {
    method: "POST",
    context,
    body: { story_id: storyId, test_plan_md: testPlanMd },
  });
}

export function getTestPlan(context: RequestContext, storyId: number) {
  return apiRequest<Phase4TestPlanResponse>(`/api/phase4/test-plan/${storyId}`, { context });
}

export function listTestPlans(context: RequestContext) {
  return apiRequest<Phase4TestPlansResponse>("/api/phase4/test-plans", { context });
}

export function deleteTestPlan(context: RequestContext, storyId: number) {
  return apiRequest<{ ok: boolean }>(`/api/phase4/test-plan/${storyId}`, {
    method: "DELETE",
    context,
  });
}

export function generateBugReport(context: RequestContext, body: Phase4GenerateBugReportRequest, signal?: AbortSignal) {
  return apiRequest<Phase4GenerateBugReportResponse>("/api/phase4/generate-bug-report", {
    method: "POST",
    context,
    body,
    timeoutMs: PHASE4_AI_TIMEOUT_MS,
    signal,
  });
}

export function passGate(
  context: RequestContext,
  storyId: number,
  scenarioResults?: Phase4ScenarioResultItem[],
) {
  return apiRequest<{ ok: boolean }>("/api/phase4/pass-gate", {
    method: "POST",
    context,
    body: {
      story_id: storyId,
      ...(scenarioResults?.length ? { scenario_results: scenarioResults } : {}),
    },
  });
}

export function failGate(context: RequestContext, body: Phase4FailGateRequest) {
  return apiRequest<{ ok: boolean }>("/api/phase4/fail-gate", {
    method: "POST",
    context,
    body,
  });
}

// ── Fix-Bolt artifacts (bug reports + fix log) ──────────────────────────────

export function listBugReports(context: RequestContext) {
  return apiRequest<Phase4BugReportsResponse>("/api/phase4/bug-reports", { context });
}

export function getBugReport(context: RequestContext, storyId: number) {
  return apiRequest<Phase4BugReportResponse>(`/api/phase4/bug-report/${storyId}`, { context });
}

export function saveBugReport(context: RequestContext, storyId: number, bugReportMd: string) {
  return apiRequest<{ ok: boolean }>("/api/phase4/save-bug-report", {
    method: "POST",
    context,
    body: { story_id: storyId, bug_report_md: bugReportMd },
  });
}

export function deleteBugReport(context: RequestContext, storyId: number) {
  return apiRequest<{ ok: boolean }>(`/api/phase4/bug-report/${storyId}`, {
    method: "DELETE",
    context,
  });
}

export function getFixLog(context: RequestContext) {
  return apiRequest<Phase4FixLogResponse>("/api/phase4/fix-log", { context });
}
